numpy/core/src/multiarray/arraytypes.c.src

```diff
@@ -4166,6 +4166,55 @@ small_correlate(const char * d_, npy_intp dstride,
     }
 }
 
+/*
+*/
+
+/* A clone function for the datetime dtype metadata */
+static NpyAuxData *
+_datetime_dtype_metadata_clone(NpyAuxData *data)
+{
+    PyArray_DatetimeDTypeMetaData *newdata =
+        (PyArray_DatetimeDTypeMetaData *)PyArray_malloc(
+                        sizeof(PyArray_DatetimeDTypeMetaData));
+    if (newdata == NULL) {
+        return NULL;
+    }
+
+    memcpy(newdata, data, sizeof(PyArray_DatetimeDTypeMetaData));
+
+    return (NpyAuxData *)newdata;
+}
+
+/*
+ * Initializes the c_metadata field
+ * must not be static, gcc 4.1.2 on redhat 5 then miscompiles this function
+ * see gh-5163
+ *
+ */
+NPY_NO_EXPORT NpyAuxData*
+_get_builtin_datetime_metadata(void)
+{
+    PyArray_DatetimeDTypeMetaData *data;
+
+    /* Allocate memory for the metadata */
+    data = PyArray_malloc(sizeof(PyArray_DatetimeDTypeMetaData));
+    if (data == NULL) {
+        return NULL;
```

Comment:
I think this path needs a `PyErr_NoMemory()`

Reply:
added